pkg/client/event_tracking.go

```diff
@@ -0,0 +1,138 @@
+package client
+
+import (
+	"encoding/hex"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// keyGenKeepTrack is used to track key generation after keep opened event is
+// received. It is used to ensure that the process execution is not duplicated,
+// e.g. when the client receives the same event multiple times. When event is
+// received, it should be noted in this struct. When the signer generation
+// process completes (no matter if it succeeded or failed), it should be removed
+// from this struct.
+type keyGenKeepTrack struct {
+	*keepEventTrack
+}
+
+// closeKeepTrack is used to track keep close events. It is used to ensure
+// that the process execution is not duplicated, e.g. when the client receives
+// the same event multiple times. When event is received, it should be noted
+// in this struct. When the signer generation process completes (no matter
+// if it succeeded or failed), it should be removed from this struct.
+type closeKeepTrack struct {
+	*keepEventTrack
+}
+
+// terminateKeepTrack is used to track keep terminate events. It is used to
+// ensure that the process execution is not duplicated, e.g. when the client
+// receives the same event multiple times. When event is received, it should be
+// noted in this struct. When the signer generation process completes (no matter
+// if it succeeded or failed), it should be removed from this struct.
+type terminateKeepTrack struct {
+	*keepEventTrack
+}
```

Comment:
Why do we need those specialized structs? Wouldn't be enough to have just three instances of `keepEventTrack` in the deduplicator?

Reply:
Good point. We do not need them, it is definitely too much. Removed in c7e3ce66e74ec2c4411d8d5189e66879dc4f2372.